paddle/fluid/imperative/reducer.h

```diff
@@ -0,0 +1,224 @@
+            "allreduce.",
+            framework::DataTypeToString(dtype)));
+    }
+  }
+
+  // context is used to select the stream for split
+  void SplitTensors(const platform::CUDADeviceContext& context) {
+    switch (dtype) {
+      case framework::proto::VarType::FP16:
+        SplitTensorsForAllReduce<platform::float16>(context, &dense_contents,
+                                                    &dense_tensors);
+        break;
+      case framework::proto::VarType::FP32:
+        SplitTensorsForAllReduce<float>(context, &dense_contents,
+                                        &dense_tensors);
+        break;
+      case framework::proto::VarType::FP64:
+        SplitTensorsForAllReduce<double>(context, &dense_contents,
+                                         &dense_tensors);
+        break;
+      default:
+        PADDLE_THROW(platform::errors::Unimplemented(
+            "Data type (%s) is not supported when it splits tensors for "
+            "allreduce.",
+            framework::DataTypeToString(dtype)));
+    }
+  }
+};
+
+struct VariableIndex {
+  // record the index in groups_
+  size_t group_index;
+  size_t inside_group_index;
+};
+
+class Reducer {
+ public:
+  explicit Reducer(
+      const std::vector<std::shared_ptr<imperative::VarBase>>& vars,
+      const std::vector<std::vector<size_t>>& group_indices,
+      const std::vector<bool>& is_sparse_gradient,
+      std::shared_ptr<imperative::ParallelContext> parallel_ctx);
+
+  virtual ~Reducer() {}
+
+  void InitializeGroups(const std::vector<std::vector<size_t>>& group_indices);
+
+  int64_t InitializeDenseGroups(const std::vector<size_t>& variable_indices_,
+                                Group* p_group);
+
+  void PrepareForBackward();
+
+  void AddDistHook(VariableWrapper* var_warpper,
+                   const VariableIndex& var_index);
+
+  void MarkVariableReady(const VariableIndex& var_index,
+                         VariableWrapper* var_warpper);
+
+  void MarkGroupReady(size_t group_index);
+
+  void FinalizeBackward();
+
+  void ReleaseReducer();
+
+  // Reducer Singleton
+  static std::shared_ptr<Reducer> SetInstance(
+      const std::vector<std::shared_ptr<imperative::VarBase>>& vars,
+      const std::vector<std::vector<size_t>>& group_indices,
+      const std::vector<bool>& is_sparse_gradient,
+      std::shared_ptr<imperative::ParallelContext> parallel_ctx) {
+    if (NULL == s_instance_) {
+      s_instance_.reset(new paddle::imperative::Reducer(
+          vars, group_indices, is_sparse_gradient, parallel_ctx));
+    }
+    return s_instance_;
+  }
+
+  static std::shared_ptr<Reducer> GetInstance() {
+    PADDLE_ENFORCE_EQ(
+        s_instance_ != NULL, true,
+        platform::errors::InvalidArgument("Reducer is not initialized."));
+    return s_instance_;
+  }
+
+ private:
+  std::vector<std::shared_ptr<imperative::VarBase>> vars_;
+  std::vector<std::vector<size_t>> group_indices_;
+  static std::shared_ptr<Reducer> s_instance_;
```

Comment:
static members may be risky
![image](https://user-images.githubusercontent.com/22561442/100352175-91a06d80-3027-11eb-9cfa-c68a7d22d1c7.png)

Reply:
This is not a standard singleton pattern. The constructor of the singleton pattern is private. Now I can new more than one `Reducer` by this design, maybe you can refer to the singleton in paddle, such as `AllocatorFacade` in paddle

Reply:
This singleton is constructed by SetInstance. It is a copy of the BoxWrapper singleton and HeterWrapper singleton in paddle.